okhttp/src/main/java/okhttp3/internal/http2/Http2Stream.java

```diff
@@ -180,9 +180,17 @@ public void sendResponseHeaders(List<Header> responseHeaders, boolean out) throw
         outFinished = true;
       }
     }
+
+    // Only DATA frames are subject to flow-control. Transmit the HEADER frame if the connection
+    // flow-control window is fully depleted.
+    boolean connectionWindowEmpty;
+    synchronized (connection) {
+      connectionWindowEmpty = connection.bytesLeftInWriteWindow == 0L;
+    }
+
     connection.writeSynReply(id, outFinished, responseHeaders);
 
-    if (outFinished) {
+    if (outFinished || connectionWindowEmpty) {
```

Comment:
Tempted to refactor this so it’s more symmetric with the code in `Http2Connection.newStream()`. In particular:
 - we could replace the `outFinished` local with a different boolean, `flushHeaders`.
 - we could only check if the connection window is empty if `flushHeaders` is false.

But as-is this definitely makes responses and requests work more similarly and therefore I think it’s a big improvement.

Reply:
ie.
```
if (!flushHeaders) {
  synchronized (connection) {
    flushHeaders = connection.bytesLeftInWriteWindow == 0L;
  }
}
```

Reply:
Turns out it isn’t quite that simple, cause we need to keep outFinished to send with the headers. Regardless, I did this fixup.